enumerable.rb

```diff
@@ -0,0 +1,129 @@
+# rubocop:disable Metrics/CyclomaticComplexity
+# rubocop:disable Metrics/PerceivedComplexity
+module Enumerable
+  def my_each
+    return to_enum(:my_each) unless block_given?
+
+    for item in self
+      yield item
+    end
+  end
+
+  def my_each_with_index
+    return to_enum(:my_each_with_index) unless block_given?
+
+    counter = 0
+    my_each do |item|
+      yield item, counter
+      counter += 1
+    end
+  end
+
+  def my_select
+    return to_enum(:my_select) unless block_given?
+
+    selected = []
+    my_each { |item| selected.push(item) if yield(item) }
+    selected
+  end
+
+  def my_all?(to_check = nil)
+    my_each { |value| return false unless value }
+
+    if block_given?
+      my_each { |value| return false unless yield(value) }
+    end
+
+    if to_check.is_a?(Class)
+      my_each { |value| return false unless value.is_a?(to_check) }
+
+    elsif to_check.is_a?(Regexp)
+      my_each { |value| return false unless value.to_s.match(to_check) }
+
+    elsif !to_check.nil?
+      my_each { |value| return false unless value == to_check }
+    end
+
+    true
+  end
+
+  def my_any?(to_check = nil)
+    if !block_given? && to_check.nil?
+      my_each { |value| return true if value }
+    end
+
+    if block_given?
+      my_each { |value| return true if yield(value) }
+    end
+
+    if to_check.is_a?(Class)
+      my_each { |value| return true if value.is_a?(to_check) }
+
+    elsif to_check.is_a?(Regexp)
+      my_each { |value| return true if value.to_s.match(to_check) }
+
+    elsif !to_check.nil?
+      my_each { |value| return true if value == to_check }
+    end
+    false
+  end
+
+  def my_none?(to_check = nil)
+    if !block_given? && to_check.nil?
+      my_each { |value| return false if value }
+    end
+
+    if block_given?
+      my_each { |value| return false if yield(value) }
+    end
+
+    if to_check.is_a?(Class)
+      my_each { |value| return false if value.is_a?(to_check) }
+
+    elsif to_check.is_a?(Regexp)
+      my_each { |value| return false if value.to_s.match(to_check) }
+
+    elsif !to_check.nil?
+      my_each { |value| return false if value == to_check }
+    end
+    true
+  end
+
+  def my_count(to_check = nil)
+    counter = 0
+    if !to_check.nil?
+      my_each { |value| counter += 1 if value == to_check }
+    elsif block_given?
+      my_each { |value| counter += 1 if yield(value) }
+    else
+      my_each { counter += 1 }
+    end
+    counter
+  end
+
+  def my_map(&my_proc)
+    return to_enum(:my_map) unless block_given?
+
+    map_list = []
+
+    my_each { |item| map_list.push(my_proc.call(item)) }
+    map_list
+  end
```

Comment:
```rb
2.6.5 :009 > array
 => [4, 5, 4, 0, 8, 4, 5, 2, 4, 5, 4, 2, 3, 4, 3, 6, 5, 0, 0, 3, 6, 6, 4, 8, 1, 4, 5, 3, 2, 2, 8, 0, 3, 8, 0, 3, 4, 4, 8, 1, 0, 7, 7, 1, 2, 8, 6, 0, 8, 7, 1, 2, 4, 6, 2, 8, 6, 2, 1, 4, 4, 0, 6, 2, 6, 4, 4, 7, 5, 2, 2, 5, 7, 1, 2, 5, 1, 5, 7, 5, 1, 3, 3, 0, 7, 2, 1, 4, 7, 2, 5, 6, 0, 2, 6, 5, 2, 1, 2, 6] 
2.6.5 :010 > my_proc = proc { |num| num > 10 }
2.6.5 :011 > array.my_map(my_proc) { |num| num < 10 }
Traceback (most recent call last):
        7: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `<main>'
        6: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/ruby_executable_hooks:24:in `eval'
        5: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `<main>'
        4: from /home/maelfosso/.rvm/gems/ruby-2.6.5/bin/irb:23:in `load'
        3: from /home/maelfosso/.rvm/gems/ruby-2.6.5/gems/irb-1.2.4/exe/irb:11:in `<top (required)>'
        2: from (irb):11
        1: from enumerable.rb:104:in `my_map'
ArgumentError (wrong number of arguments (given 1, expected 0))
2.6.5 :012 > array.map(&my_proc)
 => [false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false] 
2.6.5 :013 > 
```
it must execute only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.